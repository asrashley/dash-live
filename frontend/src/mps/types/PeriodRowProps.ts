import { MpsPeriod } from "../../types/MpsPeriod";

export interface PeriodRowProps {
  className?: string;
  index: number;
  item: MpsPeriod;
}
