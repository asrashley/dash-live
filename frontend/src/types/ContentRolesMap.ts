/* a map from role to the content types that are applicable to that role */
export type ContentRolesMap = {
    [role: string]: string[];
}