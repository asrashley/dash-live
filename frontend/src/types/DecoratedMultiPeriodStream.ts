import { MpsPeriod } from "./MpsPeriod";
import { MultiPeriodStream } from "./MultiPeriodStream";


export type DecoratedMultiPeriodStream = Omit<MultiPeriodStream, 'periods'> & {
    periods: MpsPeriod[];
    modified: boolean;
    lastModified: number;
};
