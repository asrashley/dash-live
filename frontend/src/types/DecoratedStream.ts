import { Stream } from "./Stream";
import { StreamTrack } from "./StreamTrack";


export interface DecoratedStream extends Stream {
    tracks: StreamTrack[];
}
