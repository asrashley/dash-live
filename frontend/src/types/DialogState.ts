import { DecoratedStream } from "./DecoratedStream";

export interface ConfirmDeleteDialogState {
    name: string;
    confirmed: boolean;
}

export interface MpsOptionsDialogState {
    options: object;
    lastModified: number;
    name: string;
}

export interface TrackPickerDialogState {
    pk: number | string;
    pid: string;
    guest: boolean;
    stream: DecoratedStream;
}

export interface DialogState {
    backdrop: boolean;
    confirmDelete?: ConfirmDeleteDialogState;
    mpsOptions?: MpsOptionsDialogState;
    trackPicker?: TrackPickerDialogState;
}