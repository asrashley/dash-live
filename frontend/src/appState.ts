import { createContext } from 'preact';
import { signal, computed, type Signal, type ReadonlySignal } from "@preact/signals";
import { DialogState } from './types/DialogState';
import { InitialUserState, UserState } from './types/UserState';

export interface AppStateType {
  backdrop: ReadonlySignal<boolean>;
  dialog: Signal<DialogState>;
  user: ReadonlySignal<UserState>;
  setUser: (ius: InitialUserState) => void;
}

export const AppStateContext = createContext<AppStateType>(null);

export interface CreateAppStateProps {
  userInfo: InitialUserState
}

export function createAppState(userInfo: InitialUserState = {isAuthenticated: false, groups:[]}): AppStateType {
  const dialog = signal(null);
  const userData = signal<InitialUserState>(userInfo);
  const user = computed<UserState>(() => ({
    ...userData.value,
    permissions: {
      admin: userData.value.groups.includes('ADMIN'),
      media: userData.value.groups.includes('MEDIA'),
      user: userData.value.groups.includes('USER'),
    },
  }));

  const backdrop = computed(() => {
    return dialog.value?.backdrop === true;
  });

  const setUser = (ius: InitialUserState) => {
    userData.value = {...ius};
  };

  return { dialog, backdrop, user, setUser };
}
