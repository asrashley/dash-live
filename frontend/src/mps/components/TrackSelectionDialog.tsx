import { type JSX } from "preact";
import { useCallback, useContext } from "preact/hooks";
import { Signal, useComputed } from "@preact/signals";

import { ModalDialog } from "../../components/ModalDialog";
import { MultiPeriodModelContext } from "../../hooks/useMultiPeriodStream";
import { AppStateContext } from "../../appState";
import { MpsTrack } from "../../types/MpsTrack";
import { TrackPickerDialogState } from "../../types/DialogState";
import { MultiPeriodStream } from "../../types/MultiPeriodStream";
import { ContentRolesMap } from "../../types/ContentRolesMap";
import { DecoratedMpsTrack } from "../../types/DecoratedMpsTrack";
import { useContentRoles } from "../../hooks/useContentRoles";
import { StreamTrack } from "../../types/StreamTrack";

interface RoleSelectProps {
  name: string;
  roles: Signal<string[]>;
  value: string | number;
  onChange: (ev: JSX.TargetedEvent<HTMLSelectElement>) => void;
  className: string;
  disabled?: boolean;
}

function RoleSelect({
  name,
  roles,
  value,
  onChange,
  className,
  disabled,
}: RoleSelectProps) {
  return (
    <select
      className={className}
      name={name}
      value={value}
      onChange={onChange}
      disabled={disabled}
    >
      {roles.value.map((role) => (
        <option key={role} value={role}>
          {role}
        </option>
      ))}
    </select>
  );
}

function allowedTrackRoles(
  track: MpsTrack,
  contentRoles: Signal<ContentRolesMap>
): string[] {
  const roles: string[] = [];
  for (const [name, usage] of Object.entries(contentRoles.value)) {
    if (usage.includes(track.content_type)) {
      roles.push(name);
    }
  }
  return roles;
}

interface TrackSelectRowProps {
  contentRoles: Signal<ContentRolesMap>;
  track: DecoratedMpsTrack;
  onChange: (track: DecoratedMpsTrack) => void;
  guest?: boolean;
}

function TrackSelectRow({
  contentRoles,
  track,
  onChange,
  guest,
}: TrackSelectRowProps) {
  const roles = useComputed(() => allowedTrackRoles(track, contentRoles));
  const { track_id, enabled, encrypted, clearBitrates, encryptedBitrates } =
    track;
  const bitrates = encrypted ? encryptedBitrates : clearBitrates;
  const bitratesText = bitrates > 1 ? `, ${bitrates} bitrates` : "";
  const label = `${track.content_type} track ${track_id} (${track.codec_fourcc}${bitratesText})`;

  const onToggleEnabled = () => {
    if (guest) {
      return;
    }
    onChange({
      ...track,
      enabled: !enabled,
    });
  };

  const onToggleEncrypted = () => {
    if (guest) {
      return;
    }
    onChange({
      ...track,
      encrypted: !encrypted,
    });
  };

  const onRoleChange = (ev: Event) => {
    if (guest) {
      return;
    }
    onChange({
      ...track,
      role: (ev.target as HTMLInputElement).value,
    });
  };

  return (
    <div className="input-group mb-3 row border p-1">
      <div className="col-1">
        <input
          type="checkbox"
          class="form-check-input"
          id={`id_enable_${track_id}`}
          name={`enable_${track_id}`}
          checked={enabled}
          onClick={onToggleEnabled}
          disabled={guest}
        />
      </div>
      <label className="form-check-label col-6" for={`id_enable_${track_id}`}>
        {label}
      </label>
      <RoleSelect
        roles={roles}
        value={track.role}
        className="form-select col-3"
        disabled={guest}
        name={`role_${track_id}`}
        onChange={onRoleChange}
      />
      <div className="col-1">
        <input
          className="form-check-input"
          type="checkbox"
          checked={encrypted}
          name={`enc_${track_id}`}
          onClick={onToggleEncrypted}
          disabled={guest || encryptedBitrates === 0}
        />
      </div>
    </div>
  );
}

function generateMediaTracks(
  trackPicker: Signal<TrackPickerDialogState | undefined>,
  model: Signal<MultiPeriodStream>
): DecoratedMpsTrack[] {
  if (!trackPicker.value || !model.value) {
    return [];
  }
  const { pk, stream } = trackPicker.value;
  const { periods = [] } = model.value;
  const prd = periods.find((p) => p.pk === pk);
  return stream.tracks.map((stk: StreamTrack) => {
    const trk: MpsTrack | undefined = prd?.tracks.find((t: MpsTrack) => t.track_id === stk.track_id);
    return {
      enabled: trk !== undefined,
      ...stk,
      ...trk,
    };
  });
}

export interface TrackSelectionDialogProps {
  onClose: () => void;
}

export function TrackSelectionDialog({ onClose }: TrackSelectionDialogProps) {
  const { model, modifyPeriod } = useContext(MultiPeriodModelContext);
  const { dialog } = useContext(AppStateContext);
  const { contentRoles } = useContentRoles();
  const trackPicker = useComputed<TrackPickerDialogState | undefined>(
    () => dialog.value?.trackPicker
  );
  const mediaTracks = useComputed(() =>
    generateMediaTracks(trackPicker, model)
  );
  const allSelected = useComputed<boolean>(() => {
    return !mediaTracks.value.some((trk) => !trk.enabled);
  });
  const title = useComputed<string>(
    () =>
      `${trackPicker.value?.guest ? "" : "Choose "}tracks for Period "${
        trackPicker.value?.pid
      }"`
  );

  const updateTrack = useCallback(
    (track) => {
      if (!trackPicker.value) {
        return;
      }
      modifyPeriod({
        periodPk: trackPicker.value.pk,
        track,
      });
    },
    [trackPicker.value, modifyPeriod]
  );

  const selectAllTracks = useCallback(() => {
    const newTracks = mediaTracks.value.map((trk) => ({
      ...trk,
      enabled: !allSelected.value,
    }));
    newTracks.forEach((trk) => {
      modifyPeriod({
        periodPk: trackPicker.value.pk,
        track: trk,
      });
    });
  }, [allSelected, mediaTracks, modifyPeriod, trackPicker]);

  if (!trackPicker.value) {
    return null;
  }

  const { guest = false } = trackPicker.value;

  return (
    <ModalDialog onClose={onClose} title={title.value} size="lg">
      <div className="row fw-bold">
        <div className="col-1">
          <input
            className="form-check-input"
            data-testid="select-all-tracks"
            type="checkbox"
            disabled={guest}
            name="all_tracks"
            checked={allSelected.value}
            onClick={selectAllTracks}
          />
        </div>
        <div className="col-6 text-center">Track</div>
        <div className="col-3 text-center">Role</div>
        <div className="col-2 text-end">Encrypted</div>
      </div>
      $
      {mediaTracks.value.map((trk) => (
        <TrackSelectRow
          key={trk.pk}
          track={trk}
          guest={guest}
          contentRoles={contentRoles}
          onChange={updateTrack}
        />
      ))}
    </ModalDialog>
  );
}
