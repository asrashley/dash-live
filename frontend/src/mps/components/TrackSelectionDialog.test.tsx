import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { mock } from "vitest-mock-extended";
import { signal } from "@preact/signals";
import { act, fireEvent } from "@testing-library/preact";

import { renderWithProviders } from "../../test/renderWithProviders";
import { TrackSelectionDialog } from "./TrackSelectionDialog";
import { createAppState } from "../../appState";
import {
  MpsModelValidationErrors,
  MultiPeriodModelContext,
  UseMultiPeriodStreamHook,
} from "../../hooks/useMultiPeriodStream";
import { TrackPickerDialogState } from "../../types/DialogState";
import { DecoratedStream } from "../../types/DecoratedStream";
import { ApiRequests, EndpointContext } from "../../endpoints";
import { ContentRolesMap } from "../../types/ContentRolesMap";
import { MpsPeriod } from "../../types/MpsPeriod";
import { StreamTrack } from "../../types/StreamTrack";
import { DecoratedMultiPeriodStream } from "../../types/DecoratedMultiPeriodStream";

describe("TrackSelectionDialog component", () => {
  const onClose = vi.fn();
  const appState = createAppState();
  const model = signal<DecoratedMultiPeriodStream>();
  const loaded = signal<string | undefined>();
  const modified = signal<boolean>(false);
  const errors = signal<MpsModelValidationErrors>({});
  const isValid = signal<boolean>(true);
  const mpsContext: UseMultiPeriodStreamHook = {
    model,
    errors,
    loaded,
    modified,
    isValid,
    setFields: vi.fn(),
    addPeriod: vi.fn(),
    setPeriodOrdering: vi.fn(),
    removePeriod: vi.fn(),
    modifyPeriod: vi.fn(),
    saveChanges: vi.fn(),
    deleteStream: vi.fn(),
    discardChanges: vi.fn(),
  };
  const stream: DecoratedStream = {
    tracks: [
      {
        codec_fourcc: "avc1",
        content_type: "video",
        track_id: 1,
        clearBitrates: 3,
        encryptedBitrates: 3,
      },
      {
        codec_fourcc: "mp4a",
        content_type: "audio",
        clearBitrates: 1,
        encryptedBitrates: 0,
        track_id: 2,
      },
      {
        codec_fourcc: "ec3",
        content_type: "audio",
        clearBitrates: 1,
        encryptedBitrates: 0,
        track_id: 3,
      },
    ],
    defaults: null,
    directory: "demo",
    duration: "PT9M",
    marlin_la_url: "",
    media_files: [],
    pk: 10,
    playready_la_url: "",
    timing_ref: undefined,
    title: "stream title",
  };
  const mpsPeriod: MpsPeriod = {
    pk: 5,
    duration: "",
    ordering: 0,
    parent: 0,
    pid: "pid1",
    start: "",
    stream: 10,
    tracks: [{
      ...stream.tracks[0],
      encrypted: false,
      lang: "",
      pk: 15,
      role: "main"
    },
    {
      ...stream.tracks[1],
      encrypted: false,
      lang: "",
      pk: 16,
      role: "main"
    },
    {
      ...stream.tracks[2],
      encrypted: false,
      lang: "",
      pk: 17,
      role: "alternate"
    }]
  };
  const trackPicker: TrackPickerDialogState = {
    pk: mpsPeriod.pk,
    pid: mpsPeriod.pid,
    guest: false,
    stream,
  };
  const apiRequests = mock<ApiRequests>();
  let rolesPromise: Promise<void>;

  beforeEach(() => {
    appState.dialog.value = {
      backdrop: true,
      trackPicker: { ...trackPicker },
    };
    model.value = {
      pk: 12,
      name: "test",
      title: "mps title",
      periods: [mpsPeriod],
      options: null,
      modified: false,
      lastModified: 0,
    };
    rolesPromise = new Promise<void>((resolve) => {
      apiRequests.getContentRoles.mockImplementation(async () => {
        const data = await import("../../test/fixtures/content_roles.json");
        resolve();
        return data.default as ContentRolesMap;
      });
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("should display dialog box", async () => {
    const { findByText, asFragment } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <MultiPeriodModelContext.Provider value={mpsContext}>
          <TrackSelectionDialog onClose={onClose} />
        </MultiPeriodModelContext.Provider>
      </EndpointContext.Provider>,
      { appState }
    );
    await act(async () => {
      await rolesPromise;
    });
    await findByText("Close");
    expect(asFragment()).toMatchSnapshot();
  });

  test("can close dialog box", () => {
    const { getByText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <MultiPeriodModelContext.Provider value={mpsContext}>
          <TrackSelectionDialog onClose={onClose} />
        </MultiPeriodModelContext.Provider>
      </EndpointContext.Provider>,
      { appState }
    );
    const btn = getByText("Close") as HTMLButtonElement;
    fireEvent.click(btn);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test("can select all tracks", async () => {
    const tracks = [
      mpsPeriod.tracks[0],
      mpsPeriod.tracks[1],
    ]
    const prd = {
      ...mpsPeriod,
      tracks,
    };
    model.value = {
      pk: 12,
      name: "test",
      title: "mps title",
      periods: [prd],
      options: null,
      modified: false,
      lastModified: 0,
    };

    const { getByTestId } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <MultiPeriodModelContext.Provider value={mpsContext}>
          <TrackSelectionDialog onClose={onClose} />
        </MultiPeriodModelContext.Provider>
      </EndpointContext.Provider>,
      { appState }
    );
    await act(async () => {
      await rolesPromise;
    });
    const inp = getByTestId("select-all-tracks") as HTMLInputElement;
    expect(inp.checked).toEqual(false);
    fireEvent.click(inp);
    expect(inp.checked).toEqual(true);
    expect(onClose).not.toHaveBeenCalledTimes(1);
    expect(mpsContext.modifyPeriod).toHaveBeenCalledTimes(stream.tracks.length);
    stream.tracks.forEach((trk: StreamTrack, idx: number) => {
      expect(mpsContext.modifyPeriod).toHaveBeenNthCalledWith(idx + 1, {
        periodPk: trackPicker.pk,
        track: expect.objectContaining({
          ...trk,
          enabled: true,
        }),
      });
    });
  });

  test("can toggle one track", async () => {
    const { getBySelector } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <MultiPeriodModelContext.Provider value={mpsContext}>
          <TrackSelectionDialog onClose={onClose} />
        </MultiPeriodModelContext.Provider>
      </EndpointContext.Provider>,
      { appState }
    );
    await act(async () => {
      await rolesPromise;
    });
    const inp = getBySelector("#id_enable_1") as HTMLInputElement;
    expect(inp.checked).toEqual(true);
    fireEvent.click(inp);
    expect(inp.checked).toEqual(false);
    expect(onClose).not.toHaveBeenCalledTimes(1);
    expect(mpsContext.modifyPeriod).toHaveBeenCalledTimes(1);
    expect(mpsContext.modifyPeriod).toHaveBeenCalledWith({
      periodPk: trackPicker.pk,
      track: {
        ...mpsPeriod.tracks[0],
        enabled: false,
      },
    });
  });

  test("can toggle encryption setting on one track", async () => {
    const { findBySelector } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <MultiPeriodModelContext.Provider value={mpsContext}>
          <TrackSelectionDialog onClose={onClose} />
        </MultiPeriodModelContext.Provider>
      </EndpointContext.Provider>,
      { appState }
    );
    await act(async () => {
      await rolesPromise;
    });
    const inp = await findBySelector('input[name="enc_1"]') as HTMLInputElement;
    expect(inp.checked).toEqual(false);
    fireEvent.click(inp);
    expect(inp.checked).toEqual(true);
    expect(onClose).not.toHaveBeenCalledTimes(1);
    expect(mpsContext.modifyPeriod).toHaveBeenCalledTimes(1);
    expect(mpsContext.modifyPeriod).toHaveBeenCalledWith({
      periodPk: trackPicker.pk,
      track: {
        ...mpsPeriod.tracks[0],
        encrypted: true,
        enabled: true,
      },
    });
  });

  test("can change role of a track", async () => {
    const { getBySelector } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <MultiPeriodModelContext.Provider value={mpsContext}>
          <TrackSelectionDialog onClose={onClose} />
        </MultiPeriodModelContext.Provider>
      </EndpointContext.Provider>,
      { appState }
    );
    await act(async () => {
      await rolesPromise;
    });
    const inp = getBySelector('select[name="role_2"]') as HTMLInputElement;
    expect(inp.value).toEqual("main");
    fireEvent.change(inp, { target: { value: "alternate" } });
    expect(inp.value).toEqual("alternate");
    expect(onClose).not.toHaveBeenCalledTimes(1);
    expect(mpsContext.modifyPeriod).toHaveBeenCalledTimes(1);
    expect(mpsContext.modifyPeriod).toHaveBeenCalledWith({
      periodPk: trackPicker.pk,
      track: {
        ...mpsPeriod.tracks[1],
        enabled: true,
        role: "alternate",
      },
    });
  });

  test("hide when not active", () => {
    appState.dialog.value = {
      backdrop: false,
    };
    const { container } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <MultiPeriodModelContext.Provider value={mpsContext}>
          <TrackSelectionDialog onClose={onClose} />
        </MultiPeriodModelContext.Provider>
      </EndpointContext.Provider>,
      { appState }
    );
    expect(container.innerHTML).toEqual("");
  });
});
