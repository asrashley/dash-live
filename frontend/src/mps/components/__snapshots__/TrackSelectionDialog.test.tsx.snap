// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TrackSelectionDialog component > should display dialog box 1`] = `
<DocumentFragment>
  <div
    aria-modal="true"
    class="modal fade show"
    id="dialog-box"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-lg"
      role="document"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
            Choose tracks for Period "pid1"
          </h5>
          <button
            aria-label="Close"
            class="close btn-close"
            type="button"
          />
        </div>
        <div
          class="modal-body m-2"
        >
          <div
            class="row fw-bold"
          >
            <div
              class="col-1"
            >
              <input
                class="form-check-input"
                data-testid="select-all-tracks"
                name="all_tracks"
                type="checkbox"
              />
            </div>
            <div
              class="col-6 text-center"
            >
              Track
            </div>
            <div
              class="col-3 text-center"
            >
              Role
            </div>
            <div
              class="col-2 text-end"
            >
              Encrypted
            </div>
          </div>
          $
          <div
            class="input-group mb-3 row border p-1"
          >
            <div
              class="col-1"
            >
              <input
                class="form-check-input"
                id="id_enable_1"
                name="enable_1"
                type="checkbox"
              />
            </div>
            <label
              class="form-check-label col-6"
              for="id_enable_1"
            >
              video track 1 (avc1, 3 bitrates)
            </label>
            <select
              class="form-select col-3"
              name="role_1"
            >
              <option
                value="alternate"
              >
                alternate
              </option>
              <option
                value="caption"
              >
                caption
              </option>
              <option
                value="easyreader"
              >
                easyreader
              </option>
              <option
                value="emergency"
              >
                emergency
              </option>
              <option
                value="karaoke"
              >
                karaoke
              </option>
              <option
                value="main"
              >
                main
              </option>
              <option
                value="sign"
              >
                sign
              </option>
              <option
                value="subtitle"
              >
                subtitle
              </option>
              <option
                value="supplementary"
              >
                supplementary
              </option>
            </select>
            <div
              class="col-1"
            >
              <input
                class="form-check-input"
                name="enc_1"
                type="checkbox"
              />
            </div>
          </div>
          <div
            class="input-group mb-3 row border p-1"
          >
            <div
              class="col-1"
            >
              <input
                class="form-check-input"
                id="id_enable_2"
                name="enable_2"
                type="checkbox"
              />
            </div>
            <label
              class="form-check-label col-6"
              for="id_enable_2"
            >
              audio track 2 (mp4a)
            </label>
            <select
              class="form-select col-3"
              name="role_2"
            >
              <option
                value="alternate"
              >
                alternate
              </option>
              <option
                value="commentary"
              >
                commentary
              </option>
              <option
                value="description"
              >
                description
              </option>
              <option
                value="dub"
              >
                dub
              </option>
              <option
                value="emergency"
              >
                emergency
              </option>
              <option
                value="enhanced_audio_intelligibility"
              >
                enhanced_audio_intelligibility
              </option>
              <option
                value="karaoke"
              >
                karaoke
              </option>
              <option
                value="main"
              >
                main
              </option>
              <option
                value="supplementary"
              >
                supplementary
              </option>
            </select>
            <div
              class="col-1"
            >
              <input
                class="form-check-input"
                disabled=""
                name="enc_2"
                type="checkbox"
              />
            </div>
          </div>
          <div
            class="input-group mb-3 row border p-1"
          >
            <div
              class="col-1"
            >
              <input
                class="form-check-input"
                id="id_enable_3"
                name="enable_3"
                type="checkbox"
              />
            </div>
            <label
              class="form-check-label col-6"
              for="id_enable_3"
            >
              audio track 3 (ec3)
            </label>
            <select
              class="form-select col-3"
              name="role_3"
            >
              <option
                value="alternate"
              >
                alternate
              </option>
              <option
                value="commentary"
              >
                commentary
              </option>
              <option
                value="description"
              >
                description
              </option>
              <option
                value="dub"
              >
                dub
              </option>
              <option
                value="emergency"
              >
                emergency
              </option>
              <option
                value="enhanced_audio_intelligibility"
              >
                enhanced_audio_intelligibility
              </option>
              <option
                value="karaoke"
              >
                karaoke
              </option>
              <option
                value="main"
              >
                main
              </option>
              <option
                value="supplementary"
              >
                supplementary
              </option>
            </select>
            <div
              class="col-1"
            >
              <input
                class="form-check-input"
                disabled=""
                name="enc_3"
                type="checkbox"
              />
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-secondary btn-secondary"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
